ADALiOS/ADALiOS/ADAuthenticationRequest+AcquireToken.m

```diff
@@ -0,0 +1,409 @@
+        return;
+    }
+    
+    [[ADInstanceDiscovery sharedInstance] validateAuthority:_context.authority
+                                              correlationId:_correlationId
+                                            completionBlock:^(BOOL validated, ADAuthenticationError *error)
+     {
+         if (error)
+         {
+             completionBlock([ADAuthenticationResult resultFromError:error]);
+         }
+         else
+         {
+             [self validatedAcquireToken:completionBlock];
+         }
+     }];
+
+}
+
+- (void)validatedAcquireToken:(ADAuthenticationCallback)completionBlock
+{
+    [self ensureRequest];
+    
+    //Check the cache:
+    ADAuthenticationError* error;
+    //We are explicitly creating a key first to ensure indirectly that all of the required arguments are correct.
+    //This is the safest way to guarantee it, it will raise an error, if the the any argument is not correct:
+    ADTokenCacheStoreKey* key = [ADTokenCacheStoreKey keyWithAuthority:_context.authority
+                                                              resource:_resource
+                                                              clientId:_clientId
+                                                                 error:&error];
+    if (!key)
+    {
+        //If the key cannot be extracted, call the callback with the information:
+        ADAuthenticationResult* result = [ADAuthenticationResult resultFromError:error];
+        completionBlock(result);
+        return;
+    }
+    
+    if (![ADAuthenticationContext isForcedAuthorization:_promptBehavior] && [_context hasCacheStore])
+    {
+        //Cache should be used in this case:
+        BOOL accessTokenUsable;
+        ADTokenCacheStoreItem* cacheItem = [_context findCacheItemWithKey:key userId:_identifier useAccessToken:&accessTokenUsable error:&error];
+        if (error)
+        {
+            completionBlock([ADAuthenticationResult resultFromError:error]);
+            return;
+        }
+        
+        if (cacheItem)
+        {
+            //Found a promising item in the cache, try using it:
+            [self attemptToUseCacheItem:cacheItem
+                         useAccessToken:accessTokenUsable
+                        completionBlock:completionBlock];
+            return; //The tryRefreshingFromCacheItem has taken care of the token obtaining
+        }
+    }
+    
+    [self requestToken:completionBlock];
+}
+
+/*Attemps to use the cache. Returns YES if an attempt was successful or if an
+ internal asynchronous call will proceed the processing. */
+- (void)attemptToUseCacheItem:(ADTokenCacheStoreItem*)item
+               useAccessToken:(BOOL)useAccessToken
+              completionBlock:(ADAuthenticationCallback)completionBlock
+{
+    //All of these should be set before calling this method:
+    THROW_ON_NIL_ARGUMENT(completionBlock);
+    AD_REQUEST_CHECK_ARGUMENT(item);
+    AD_REQUEST_CHECK_PROPERTY(_resource);
```

Comment:
_resource and _clientid are related to request itself not to this specific method call. If you check those when creating a request, it will reduce similar checks.